### PR TITLE
[Cisco Duo] New data stream for Activity logs

### PR DESCRIPTION
**Pull request by chemamartinez**
## Proposed commit message

Added new data stream `activity` to collect [Activity logs](https://duo.com/docs/adminapi#activity) from Cisco Duo.

The CEL program follows the API specifications set at:
  - [Authentication](https://duo.com/docs/adminapi#authentication)
  - [Response format](https://duo.com/docs/adminapi#response-format)
  - [Parameters and pagination](https://duo.com/docs/adminapi#activity)

It also adds a feature request for the `auth` data stream, including geo enrichment for the IP into the `access_device` field.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
- [x] I have verified that any added dashboard complies with Kibana's [Dashboard good practices](https://docs.elastic.dev/ux-guidelines/data-viz/dashboard-good-practices) 

## How to test this PR locally

Added pipeline and system tests for the new data stream:
```
Run asset tests for the package
--- Test results for package: cisco_duo - START ---
╭───────────┬────────────────────┬───────────┬────────────────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE   │ DATA STREAM        │ TEST TYPE │ TEST NAME                                                          │ RESULT │ TIME ELAPSED │
├───────────┼────────────────────┼───────────┼────────────────────────────────────────────────────────────────────┼────────┼──────────────┤
│ cisco_duo │                    │ asset     │ dashboard cisco_duo-0607d4a3-5322-41c1-b8fa-f0d29bcc2757 is loaded │ PASS   │      1.584µs │
│ cisco_duo │                    │ asset     │ dashboard cisco_duo-118061c3-3130-4c71-8899-157bab1ad447 is loaded │ PASS   │        292ns │
│ cisco_duo │                    │ asset     │ dashboard cisco_duo-5a0b80af-49ad-42ee-89b7-c89faa927826 is loaded │ PASS   │        416ns │
│ cisco_duo │                    │ asset     │ dashboard cisco_duo-7a135061-78a3-45d9-951b-4b9b665fa729 is loaded │ PASS   │        208ns │
│ cisco_duo │                    │ asset     │ dashboard cisco_duo-b386f94c-0856-4508-ba08-a525a2f3b70f is loaded │ PASS   │        417ns │
│ cisco_duo │                    │ asset     │ dashboard cisco_duo-c3336a66-68ff-4bcd-95ff-fb388793f721 is loaded │ PASS   │        375ns │
│ cisco_duo │                    │ asset     │ dashboard cisco_duo-e91470e5-2ded-4ff1-8bb5-24e06b949c1d is loaded │ PASS   │        250ns │
│ cisco_duo │                    │ asset     │ map cisco_duo-158c0e80-148c-11ec-9386-31989719f9db is loaded       │ PASS   │        250ns │
│ cisco_duo │                    │ asset     │ search cisco_duo-64869c89-8c44-4644-a84a-9815c0fddba0 is loaded    │ PASS   │        375ns │
│ cisco_duo │ activity           │ asset     │ index_template logs-cisco_duo.activity is loaded                   │ PASS   │        250ns │
│ cisco_duo │ activity           │ asset     │ ingest_pipeline logs-cisco_duo.activity-2.2.0 is loaded            │ PASS   │        292ns │
│ cisco_duo │ admin              │ asset     │ index_template logs-cisco_duo.admin is loaded                      │ PASS   │        166ns │
│ cisco_duo │ admin              │ asset     │ ingest_pipeline logs-cisco_duo.admin-2.2.0 is loaded               │ PASS   │        292ns │
│ cisco_duo │ auth               │ asset     │ index_template logs-cisco_duo.auth is loaded                       │ PASS   │        334ns │
│ cisco_duo │ auth               │ asset     │ ingest_pipeline logs-cisco_duo.auth-2.2.0 is loaded                │ PASS   │        125ns │
│ cisco_duo │ offline_enrollment │ asset     │ index_template logs-cisco_duo.offline_enrollment is loaded         │ PASS   │        209ns │
│ cisco_duo │ offline_enrollment │ asset     │ ingest_pipeline logs-cisco_duo.offline_enrollment-2.2.0 is loaded  │ PASS   │        125ns │
│ cisco_duo │ summary            │ asset     │ index_template logs-cisco_duo.summary is loaded                    │ PASS   │        208ns │
│ cisco_duo │ summary            │ asset     │ ingest_pipeline logs-cisco_duo.summary-2.2.0 is loaded             │ PASS   │        125ns │
│ cisco_duo │ telephony          │ asset     │ index_template logs-cisco_duo.telephony is loaded                  │ PASS   │        209ns │
│ cisco_duo │ telephony          │ asset     │ ingest_pipeline logs-cisco_duo.telephony-2.2.0 is loaded           │ PASS   │        125ns │
│ cisco_duo │ telephony_v2       │ asset     │ index_template logs-cisco_duo.telephony_v2 is loaded               │ PASS   │        375ns │
│ cisco_duo │ telephony_v2       │ asset     │ ingest_pipeline logs-cisco_duo.telephony_v2-2.2.0 is loaded        │ PASS   │        166ns │
│ cisco_duo │ trust_monitor      │ asset     │ index_template logs-cisco_duo.trust_monitor is loaded              │ PASS   │        375ns │
│ cisco_duo │ trust_monitor      │ asset     │ ingest_pipeline logs-cisco_duo.trust_monitor-2.2.0 is loaded       │ PASS   │        125ns │
╰───────────┴────────────────────┴───────────┴────────────────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: cisco_duo - END   ---
Done
Run pipeline tests for the package
--- Test results for package: cisco_duo - START ---
╭───────────┬────────────────────┬───────────┬────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE   │ DATA STREAM        │ TEST TYPE │ TEST NAME                                              │ RESULT │ TIME ELAPSED │
├───────────┼────────────────────┼───────────┼────────────────────────────────────────────────────────┼────────┼──────────────┤
│ cisco_duo │ activity           │ pipeline  │ (ingest pipeline warnings test-activity.log)           │ PASS   │ 196.669625ms │
│ cisco_duo │ activity           │ pipeline  │ test-activity.log                                      │ PASS   │ 134.559833ms │
│ cisco_duo │ admin              │ pipeline  │ (ingest pipeline warnings test-admin.log)              │ PASS   │ 206.523291ms │
│ cisco_duo │ admin              │ pipeline  │ (ingest pipeline warnings test-empty.log)              │ PASS   │ 202.109375ms │
│ cisco_duo │ admin              │ pipeline  │ test-admin.log                                         │ PASS   │    151.045ms │
│ cisco_duo │ admin              │ pipeline  │ test-empty.log                                         │ PASS   │  40.457458ms │
│ cisco_duo │ auth               │ pipeline  │ (ingest pipeline warnings test-auth.log)               │ PASS   │ 218.412834ms │
│ cisco_duo │ auth               │ pipeline  │ (ingest pipeline warnings test-empty.log)              │ PASS   │   194.9385ms │
│ cisco_duo │ auth               │ pipeline  │ test-auth.log                                          │ PASS   │ 345.084917ms │
│ cisco_duo │ auth               │ pipeline  │ test-empty.log                                         │ PASS   │   40.33725ms │
│ cisco_duo │ offline_enrollment │ pipeline  │ (ingest pipeline warnings test-empty.log)              │ PASS   │  198.41075ms │
│ cisco_duo │ offline_enrollment │ pipeline  │ (ingest pipeline warnings test-offline-enrollment.log) │ PASS   │   200.7585ms │
│ cisco_duo │ offline_enrollment │ pipeline  │ test-empty.log                                         │ PASS   │  37.008042ms │
│ cisco_duo │ offline_enrollment │ pipeline  │ test-offline-enrollment.log                            │ PASS   │  44.096708ms │
│ cisco_duo │ summary            │ pipeline  │ (ingest pipeline warnings test-summary.log)            │ PASS   │ 205.246834ms │
│ cisco_duo │ summary            │ pipeline  │ test-summary.log                                       │ PASS   │   41.75075ms │
│ cisco_duo │ telephony          │ pipeline  │ (ingest pipeline warnings test-empty.log)              │ PASS   │ 234.215875ms │
│ cisco_duo │ telephony          │ pipeline  │ (ingest pipeline warnings test-telephony.log)          │ PASS   │ 247.474375ms │
│ cisco_duo │ telephony          │ pipeline  │ test-empty.log                                         │ PASS   │  40.395042ms │
│ cisco_duo │ telephony          │ pipeline  │ test-telephony.log                                     │ PASS   │  38.864542ms │
│ cisco_duo │ telephony_v2       │ pipeline  │ (ingest pipeline warnings test-telephony-v2.log)       │ PASS   │ 210.205625ms │
│ cisco_duo │ telephony_v2       │ pipeline  │ test-telephony-v2.log                                  │ PASS   │   43.08375ms │
│ cisco_duo │ trust_monitor      │ pipeline  │ (ingest pipeline warnings test-trust-monitor.log)      │ PASS   │ 204.730667ms │
│ cisco_duo │ trust_monitor      │ pipeline  │ test-trust-monitor.log                                 │ PASS   │  49.444791ms │
╰───────────┴────────────────────┴───────────┴────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: cisco_duo - END   ---
Done
Run policy tests for the package
--- Test results for package: cisco_duo - START ---
No test results
--- Test results for package: cisco_duo - END   ---
Done
Run static tests for the package
--- Test results for package: cisco_duo - START ---
╭───────────┬────────────────────┬───────────┬──────────────────────────┬────────┬──────────────╮
│ PACKAGE   │ DATA STREAM        │ TEST TYPE │ TEST NAME                │ RESULT │ TIME ELAPSED │
├───────────┼────────────────────┼───────────┼──────────────────────────┼────────┼──────────────┤
│ cisco_duo │ activity           │ static    │ Verify sample_event.json │ PASS   │  52.114292ms │
│ cisco_duo │ admin              │ static    │ Verify sample_event.json │ PASS   │  41.618833ms │
│ cisco_duo │ auth               │ static    │ Verify sample_event.json │ PASS   │  57.805375ms │
│ cisco_duo │ offline_enrollment │ static    │ Verify sample_event.json │ PASS   │  42.232042ms │
│ cisco_duo │ summary            │ static    │ Verify sample_event.json │ PASS   │  38.962625ms │
│ cisco_duo │ telephony          │ static    │ Verify sample_event.json │ PASS   │  38.040167ms │
│ cisco_duo │ telephony_v2       │ static    │ Verify sample_event.json │ PASS   │  37.535666ms │
│ cisco_duo │ trust_monitor      │ static    │ Verify sample_event.json │ PASS   │  42.190791ms │
╰───────────┴────────────────────┴───────────┴──────────────────────────┴────────┴──────────────╯
--- Test results for package: cisco_duo - END   ---
Done
Run system tests for the package
--- Test results for package: cisco_duo - START ---
╭───────────┬────────────────────┬───────────┬───────────┬────────┬───────────────╮
│ PACKAGE   │ DATA STREAM        │ TEST TYPE │ TEST NAME │ RESULT │  TIME ELAPSED │
├───────────┼────────────────────┼───────────┼───────────┼────────┼───────────────┤
│ cisco_duo │ activity           │ system    │ default   │ PASS   │ 36.870015625s │
│ cisco_duo │ admin              │ system    │ default   │ PASS   │  33.85713475s │
│ cisco_duo │ auth               │ system    │ default   │ PASS   │ 33.772624833s │
│ cisco_duo │ offline_enrollment │ system    │ default   │ PASS   │ 38.540788792s │
│ cisco_duo │ summary            │ system    │ default   │ PASS   │ 34.455289708s │
│ cisco_duo │ telephony          │ system    │ default   │ PASS   │  33.41083725s │
│ cisco_duo │ telephony_v2       │ system    │ default   │ PASS   │ 39.567072958s │
│ cisco_duo │ trust_monitor      │ system    │ default   │ PASS   │ 39.527218083s │
╰───────────┴────────────────────┴───────────┴───────────┴────────┴───────────────╯
--- Test results for package: cisco_duo - END   ---
Done
```

## Related issues

- Closes https://github.com/elastic/integrations/issues/10960
- Closes https://github.com/elastic/integrations/issues/11333

## Screenshots

**Data stream configuration**

Screenshot 2024-10-10 at 19 53 26

**Dashboard**

dashboard-activity

### CODE REVIEW

File: packages/cisco_duo/data_stream/activity/_dev/test/system/test-default-config.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+vars:
+  hostname: http://{{Hostname}}:{{Port}}
+  secret_key: 40_characters_long_secret_key
+  integration_key: temp_integration_key
+  enable_request_tracer: true
+data_stream:
+  vars:
+    preserve_original_event: true
+assert:
+  hit_count: 1
```
Comment (efd6) [inline packages/cisco_duo/data_stream/activity/_dev/test/system/test-default-config.yml:12]: Can we make the test go through at least one round of pagination?

File: packages/cisco_duo/data_stream/activity/agent/stream/cel.yml.hbs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                            "message": item.encode_json(),
+                        }
+                    ),
+                    "url": state.url,
+                    "integration_key": state.integration_key,
+                    "secret_key": state.secret_key,
+                    "limit": state.limit,
+                    "mintime": state.mintime,
+                    "maxtime": state.maxtime,
+                    "date": now.format(time_layout.RFC1123Z),
```
Comment (efd6) [inline packages/cisco_duo/data_stream/activity/agent/stream/cel.yml.hbs:90]: This is already defined as `state.date` from the preamble.

File: packages/cisco_duo/data_stream/activity/agent/stream/cel.yml.hbs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                    "want_more": has(body.?response.?metadata.next_offset),
+                    ?"next_offset": (has(body.?response.?metadata.next_offset) && body.response.metadata.next_offset != null) ?
+                                        optional.of(string(body.response.metadata.next_offset))
+                                    :
+                                        optional.none(),
+                    "next_url": (has(body.?response.?metadata.next_offset) && body.response.metadata.next_offset != null) ?
```
Comment (efd6) [inline packages/cisco_duo/data_stream/activity/agent/stream/cel.yml.hbs:96]:
```suggestion
                    "next_url": (body.?response.?metadata.next_offset.orValue(null) != null) ?
```

File: packages/cisco_duo/data_stream/activity/agent/stream/cel.yml.hbs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                                "maxtime": [string(int(state.maxtime))],
+                                "mintime": [string(int(state.mintime))],
+                                "next_offset": [string(body.response.metadata.next_offset)],
+                                "sort": ["ts:asc"],
+                            }.format_query()
+                        )
+                    :
+                        state.url,
+                    "cursor": {
+                        ?"last_published": (has(body.?response.?metadata.next_offset) && body.response.metadata.next_offset != null) ?
```
Comment (efd6) [inline packages/cisco_duo/data_stream/activity/agent/stream/cel.yml.hbs:109]:
```suggestion
                        ?"last_published": (body.?response.?metadata.next_offset.orValue(null) != null) ?
```

File: packages/cisco_duo/data_stream/activity/elasticsearch/ingest_pipeline/default.yml [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        - set:
+            field: event.kind
+            value: pipeline_error
+        - append:
+            field: error.message
+            value: |-
+              Processor "{{{ _ingest.on_failure_processor_type }}}" with tag "{{{ _ingest.on_failure_processor_tag }}}" in pipeline "{{{ _ingest.on_failure_pipeline }}}" failed with message "{{{ _ingest.on_failure_message }}}"
+  - drop:
+      description: Drop if no timestamp (invalid json)
+      if: ctx.json?.ts == null
```
Comment (efd6) [inline packages/cisco_duo/data_stream/activity/elasticsearch/ingest_pipeline/default.yml:29]: Same [comment](https://github.com/elastic/integrations/pull/11327#discussion_r1796087970) as for trust monitor.

File: packages/cisco_duo/data_stream/activity/elasticsearch/ingest_pipeline/default.yml [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+  - fingerprint:
+      tag: fingerprint_activity
+      fields:
+        - json.activity_id
+        - json.akey
+        - json.ts
+      target_field: _id
+      ignore_missing: true
+
+  ## Custom fields
```
Comment (efd6) [inline packages/cisco_duo/data_stream/activity/elasticsearch/ingest_pipeline/default.yml:48]: tags

File: packages/cisco_duo/data_stream/activity/elasticsearch/ingest_pipeline/default.yml [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+      copy_from: cisco_duo.activity.id
+      ignore_empty_value: true
+  - set:
+      field: event.action
+      copy_from: cisco_duo.activity.action.name
+      ignore_empty_value: true
+  - set:
+      field: event.outcome
+      value: success
+      if: ctx.cisco_duo?.activity?.outcome != null && ctx.cisco_duo.activity.outcome == 'SUCCESS'
```
Comment (efd6) [inline packages/cisco_duo/data_stream/activity/elasticsearch/ingest_pipeline/default.yml:135]:
```suggestion
      if: ctx.cisco_duo?.activity?.outcome == 'SUCCESS'
```

File: packages/cisco_duo/data_stream/activity/elasticsearch/ingest_pipeline/default.yml [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+  - set:
+      field: event.outcome
+      value: failure
+      if: ctx.cisco_duo?.activity?.outcome != null && ctx.cisco_duo.activity.outcome == 'FAILURE'
```
Comment (efd6) [inline packages/cisco_duo/data_stream/activity/elasticsearch/ingest_pipeline/default.yml:139]:
```suggestion
      if: ctx.cisco_duo?.activity?.outcome == 'FAILURE'
```

File: packages/cisco_duo/data_stream/activity/elasticsearch/ingest_pipeline/default.yml [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+      copy_from: cisco_duo.activity.access_device.geo
+      ignore_empty_value: true
+  - set:
+      field: source.as
+      copy_from: cisco_duo.activity.access_device.as
+      ignore_empty_value: true
+  - set:
+      field: user.name
+      copy_from: cisco_duo.activity.actor.name
+      if: ctx.cisco_duo?.activity?.actor?.type != null && ctx.cisco_duo.activity.actor.type == 'user'
```
Comment (efd6) [inline packages/cisco_duo/data_stream/activity/elasticsearch/ingest_pipeline/default.yml:155]:
```suggestion
      if: ctx.cisco_duo?.activity?.actor?.type == 'user'
```

**Review by efd6 — APPROVED** [review]
Though note the concern about the string(list) conversion.

**Review by efd6 — CHANGES_REQUESTED** [review]
Something about real-world Cisco Duo documents is causing breakage with this code (as it exists in auth). So blocking.

File: packages/cisco_duo/data_stream/trust_monitor/agent/stream/cel.yml.hbs [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+                    "events": {
+                        "error": {
+                            "code": has(body.code) ? string(body.code) : string(resp.StatusCode),
+                            "id": string(resp.Status),
+                            "message": "GET: " +
+                            (
+                                (has(body.message) && body.message != "") ?
+                                    string(body.message) +
+                                        (has(body.message_detail) ?
+                                            ": " + string(body.message_detail)
```
[the remark this answers is not in the document]
Comment (chemamartinez) [reply packages/cisco_duo/data_stream/trust_monitor/agent/stream/cel.yml.hbs]: Resolving this comment as it was caused by `next_offset` and is already solved https://github.com/elastic/integrations/pull/11422

File: packages/cisco_duo/data_stream/activity/agent/stream/cel.yml.hbs [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+                    "integration_key": state.integration_key,
+                    "secret_key": state.secret_key,
+                    "limit": state.limit,
+                    "mintime": state.mintime,
+                    "maxtime": state.maxtime,
+                    "want_more": has(body.?response.?metadata.next_offset),
+                    ?"next_offset": (body.?response.?metadata.next_offset.orValue(null) != null) ?
+                                        optional.of(string(body.response.metadata.next_offset))
+                                    :
+                                        optional.none(),
```
Comment (efd6) [inline packages/cisco_duo/data_stream/activity/agent/stream/cel.yml.hbs:96]: Same comment here as [this](https://github.com/elastic/integrations/pull/11327#discussion_r1807937159).

**Review by efd6 — APPROVED** [review]
LGTM after nit addressed.

**Pull request merged**